crates/syn2mas/src/mas_writer/mod.rs

```diff
@@ -667,5 +667,185 @@ impl<'writer, 'conn> MasUserWriteBuffer<'writer, 'conn> {
+    #[serde(transparent)]
+    struct TableSnapshot {
+        rows: BTreeSet<RowSnapshot>,
+    }
+
+    #[derive(PartialEq, Eq, PartialOrd, Ord, Serialize)]
+    #[serde(transparent)]
+    struct RowSnapshot {
+        columns_to_values: BTreeMap<String, Option<String>>,
+    }
+
+    const SKIPPED_TABLES: &[&str] = &["_sqlx_migrations"];
+
+    /// Produces a serialisable snapshot of a database, usable for snapshot testing
+    ///
+    /// For brevity, empty tables, as well as [`SKIPPED_TABLES`], will not be included in the snapshot.
+    async fn snapshot_database(conn: &mut PgConnection) -> DatabaseSnapshot {
+        let mut out = DatabaseSnapshot::default();
+        let table_names: Vec<String> = sqlx::query_scalar(
+            "SELECT table_name FROM information_schema.tables WHERE table_schema = current_schema();",
+        )
+        .fetch_all(&mut *conn)
+        .await
+        .unwrap();
+
+        for table_name in table_names {
+            if SKIPPED_TABLES.contains(&table_name.as_str()) {
+                continue;
+            }
+
+            let column_names: Vec<String> = sqlx::query_scalar(
+                "SELECT column_name FROM information_schema.columns WHERE table_name = $1 AND table_schema = current_schema();"
+            ).bind(&table_name).fetch_all(&mut *conn).await.expect("failed to get column names for table for snapshotting");
+
+            let column_name_list = column_names
+                .iter()
+                // stringify all the values for simplicity
+                .map(|column_name| format!("{column_name}::TEXT AS \"{column_name}\""))
+                .collect::<Vec<_>>()
+                .join(", ");
+
+            let table_rows = sqlx::query(&format!("SELECT {column_name_list} FROM {table_name};"))
+                .fetch(&mut *conn)
+                .map_ok(|row| {
+                    let mut columns_to_values = BTreeMap::new();
+                    for (idx, column) in row.columns().iter().enumerate() {
+                        columns_to_values.insert(column.name().to_owned(), row.get(idx));
+                    }
+                    RowSnapshot { columns_to_values }
+                })
+                .try_collect::<BTreeSet<RowSnapshot>>()
+                .await
+                .expect("failed to fetch rows from table for snapshotting");
+
+            if !table_rows.is_empty() {
+                out.tables
+                    .insert(table_name, TableSnapshot { rows: table_rows });
+            }
+        }
+
+        out
+    }
+
+    /// Make a snapshot assertion against the database.
+    macro_rules! assert_db_snapshot {
+        ($db: expr) => {
+            let db_snapshot = snapshot_database($db).await;
+            ::insta::assert_yaml_snapshot!(db_snapshot);
+        };
+    }
+
+    /// Runs some code with a `MasWriter`.
+    ///
+    /// The callback is responsible for `finish`ing the `MasWriter`.
+    async fn with_mas_writer<F, R>(pool: &PgPool, callback: F) -> R
```

Comment:
Couldn't that be simplified with a function which initialises and returns a `MasWriter`? or is it because it borrows the `main_conn`?
In which case you could solve this by passing the main connection in arguments?

Reply:
good point, this started off because I needed to make a few conns but then afterwards I needed a connection to do the snapshot anyway, so in the new commit I just use that same connection as the MasWriter connection. That is quite a bit nicer (though 90% of that is because async closures are way more painful than they should be, for understandable reasons but still..)